vent/coordinator/coordinator.py

```diff
@@ -72,127 +121,123 @@ def __init__(self, sim_mode=False):
+            with self.lock:
+                self.COPY_sensor_values = sensor_values
+                self.last_message_timestamp = sensor_values.timestamp
+            for name in values.controllable_values:
+                with self.lock:
+                    not_in_control_settings = name not in self.COPY_control_settings
+                if not_in_control_settings:
                     control_setting = self.control_module.get_control(name)
-                    self.lock.acquire()
-                    self.COPY_control_settings[name] = control_setting
-                    self.lock.release()
-
-                self.lock.acquire()
-                disagreed_tentative = name in self.COPY_tentative_control_settings and self.COPY_tentative_control_settings[name] != self.COPY_control_settings[
-                            name]
-                self.lock.release()
+                    with self.lock:
+                        self.COPY_control_settings[name] = control_setting
+                with self.lock:
+                    disagreed_tentative = name in self.COPY_tentative_control_settings and \
+                                      self.COPY_tentative_control_settings[name] != self.COPY_control_settings[
+                                          name]
                 if disagreed_tentative:
-                    self.lock.acquire()
-                    tentative_control_setting = self.COPY_tentative_control_settings[name]
-                    self.lock.release()
+                    with self.lock:
+                        tentative_control_setting = self.COPY_tentative_control_settings[name]
                     self.control_module.set_control(tentative_control_setting)
-                    self.lock.acquire()
-                    self.COPY_control_settings[name] = self.control_module.get_control(name)
-                    self.lock.release()
+                    with self.lock:
+                        self.COPY_control_settings[name] = self.control_module.get_control(name)
             # sleep 10 ms
             time.sleep(0.01)
 
-    def stop(self):
-        self.stopping.set()
-
 
 class CoordinatorRemote(CoordinatorBase):
     def __init__(self, sim_mode=False):
         super().__init__(sim_mode=sim_mode)
-        # TODO: pass max_heartbeat_interval
-        self.process_manager = ProcessManager()
-        self.rpc = IPC()
-        raise NotImplementedError
+        # TODO: according to documentation, pass max_heartbeat_interval
+        self.ipc = IPC(listen=True)
+        self.process_manager = ProcessManager(sim_mode, self.ipc.addr, self.ipc.port)
+        # TODO: make sure the ipc connection is setup. There should be a clever method
+        time.sleep(1)
+        self.receive_thread = threading.Thread(target=self.__receive_loop, daemon=True)
+        self.receive_thread.start()
+        self.receive_thread_id = self.receive_thread.ident
+        self.send_thread = threading.Thread(target=self.__send_loop, daemon=True)
```

Comment:
Can send and receive be in the same thread / loop? i.e. we send a message to the controller (get_sensors, set_controls) and then wait for a reply with a timeout?

Reply:
Changing to rpc solve these issues.